src/internal/feature/feature.go

```diff
@@ -0,0 +1,224 @@
+// match, User feature state takes precedence.
+type Enabled bool
+
+// Since marks the Zarf version that a feature is first released in Alpha. By convention Since is semver starting with
+// "v": "v1.0.0".
+type Since string
+
+// Until marks the intended Zarf version that a Deprecated feature will or has been fully removed. If Deprecation must
+// be delayed, then this version be updated. By convention Until is semver starting with "v": "v1.0.0".
+type Until string
+
+// Stage describes the lifecycle of a feature, as well as its production-readiness.
+type Stage string
+
+var (
+	// Alpha features are experimental and are highly subject to change.
+	Alpha Stage = "alpha"
+	// Beta features have solidified into a release candidate and are ready for user feedback and trialed with realistic
+	// workflows. Beta features can still change before release but should be considered nearly complete.
+	Beta Stage = "beta"
+	// GA features have been fully released and are available for production usage.
+	GA Stage = "ga"
+	// Deprecated features wrap functionality that is intended to be removed in a future version. They will start as
+	// enabled by default with a warning, and eventually be disabled by default with the Until field documenting when
+	// the feature is expected to be fully removed. Disabled Deprecated feature structs should be kept in Defaults to
+	// aid in documentation.
+	Deprecated Stage = "deprecated"
+)
+
+// Feature models a Default or User-configured feature flag and its metadata.
+type Feature struct {
+	// Name stores the name of the feature flag.
+	Name `json:"name,omitempty"`
+	// Description describes how the flag is used.
+	Description `json:"description,omitempty"`
+	// Enabled describes whether a feature is explicitly enabled or disabled. A feature that does not exist in any set
+	// is considered disabled.
+	Enabled `json:"enabled,omitempty"`
+	// Since is the version a feature is first introduced in alpha stage.
+	Since `json:"since,omitempty"`
+	// Until is the version when a deprecated feature is fully removed. Historical versions included.
+	Until `json:"until,omitempty"`
+	Stage `json:"stage,omitempty"`
+}
+
+// IsEnabled allows users to optimistically check for a feature. Useful for control flow. Any user-enabled or disabled
+// features take precedence over the default setting.
+func IsEnabled(name Name) bool {
+	f, err := Get(name)
+	if err != nil {
+		// We don't actually need to check the error here because an empty f will have the same truthiness
+		return false
+	}
+	return bool(f.Enabled)
+}
+
+// Set takes a slice of one or many flags, inserting the features onto user-configured features. If a feature name is
+// provided that is already a part of the set, then Set will return an error.
+// TODO: Should we allow users to call this multiple times even if we don't allow them to overwrite features?
```

Comment:
I think dis-allowing is fine. If there's a world where we need to allow multiple sets in the future then we could change it, but this is probably the most intuitive option since there's always an error on the second attempt